src/rulesets/Base/mapreduce.jl

```diff
@@ -322,3 +325,122 @@ end
     end
     return dx
 end
+
+#####
+##### `foldl`
+#####
+
+# `foldl` guarantees to execute `f` in order, left to right. So it makes sense even when
+# this `f` is stateful, in which case the gradient must be calculated in the reverse order. 
+# The implementation aims to be efficient for both tuples and arrays. 
+# Note that it does not return a gradient for `init`.
+
+function rrule(
+        config::RuleConfig{>:HasReverseMode}, ::typeof(foldl), f::F, x::Union{AbstractArray, Tuple};
+        init=_InitialValue()
+    ) where {F}
+    list, start = if init !== _InitialValue()
+        x, init
+    else
+        _drop1(x), first(x)
+    end
+    hobbits = accumulate(list; init=(start, pi)) do (a,_), b
+        c, back = rrule_via_ad(config, f, a, b)  # LHS is just documentation here
+        # Don't need to store every `c`, need last + one for the next iteration
+    end
+    y = first(last(hobbits))
+    axe = x isa AbstractArray ? axes(x) : nothing
+    type = Val(typeof(x))
+    project = x isa AbstractArray{<:Number} ? ProjectTo(x) : identity
+    function unfoldl(dy)
+        _tini = (pi, unthunk(dy), pi)
+        trio = accumulate(_reverse1(hobbits); init=_tini) do (_,dc,_), (_,back)
+            ds, da, db = back(dc)  # 's' for self, don't need to write LHS
+            # Don't need to store every `da`, need one for the next iteration + maybe last
+        end
+        df = sum(first, trio)
+        dx = map(t -> t[3], _reverse1(trio))
+        if init === _InitialValue()
+            # `hobbits` is one short
+            dx = _vcat1(trio[end][2], dx)
+        end
+        dx_nice = axe isa Tuple ? reshape(dx, axe) : Tangent{_val(type)}(dx...)
+        return (NoTangent(), df, project(dx_nice))
+    end
+    return y, unfoldl
+end
+
+if VERSION >= v"1.6"
+    _reverse1(x) = Iterators.reverse(x)
+    _drop1(x) = Iterators.drop(x, 1)
+    _zip2(x, y) = zip(x, y)  # for `accumulate`, below
+else
+    # Old versions don't support accumulate(::itr), nor multi-dim reverse
+    _reverse1(x) = reverse(vec(x))
+    _drop1(x) = vec(x)[1:end-1]
+    _zip2(x, y) = collect(zip(x, y))
+end
+_reverse1(x::Tuple) = reverse(x)
+_drop1(x::Tuple) = Base.tail(x)
+_zip2(x::Tuple{Vararg{Any,N}}, y::Tuple{Vararg{Any,N}}) where N = ntuple(i -> (x[i],y[i]), N)
+
+struct _InitialValue end  # Old versions don't have Base._InitialValue
+
+_vcat1(x, ys::AbstractVector) = vcat(x, ys)
+_vcat1(x::AbstractArray, ys::AbstractVector) = vcat([x], ys)
+_vcat1(x, ys::Tuple) = (x, ys...)
+
+_no_tuple_tangent(dx::Tangent) = ChainRulesCore.backing(dx)
+_no_tuple_tangent(dx) = dx
```

Comment:
This is a larger zoo of special internal functions than ideal. 

Base has `_reverse` which is eager on tuples but lazy on arrays, which it uses in e.g. `foldr`; this `_reverse1` is a way to allow for older Julia versions.

Maybe there are pre-existing substitutes for some of the others?

Reply:
Can we have little 1 liner docstrings for each of these?

Reply:
Can do, but will wait until sure they survive. 

Should `_no_tuple_tangent` be something which `ChainRulesCore` provides?

Reply:
Let's open an issue about it